presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java

```diff
@@ -1209,6 +1211,219 @@ private void assertPushdownSubscripts(String tableName)
                 ImmutableMap.of("z", toSubfields("z[1][2]")));
     }
 
+    @Test
+    public void testPushDownSubfieldsFromLambdas()
+    {
+        final String tableName = "test_pushdown_subfields_from_array_lambda";
+        assertUpdate("CREATE TABLE " + tableName + "(id bigint, " +
+                "a array(bigint), " +
+                "b array(array(varchar)), " +
+                "mi map(int,array(row(a1 bigint, a2 double))), " +
+                "mv map(varchar,array(row(a1 bigint, a2 double))), " +
+                "m1 map(int,row(a1 bigint, a2 double)), " +
+                "m2 map(int,row(a1 bigint, a2 double)), " +
+                "m3 map(int,row(a1 bigint, a2 double)), " +
+                "r row(a array(row(a1 bigint, a2 double)), i bigint, d row(d1 bigint, d2 double)), " +
+                "y array(row(a bigint, b varchar, c double, d row(d1 bigint, d2 double))), " +
+                "yy array(row(a bigint, b varchar, c double, d row(d1 bigint, d2 double))), " +
+                "yyy array(row(a bigint, b varchar, c double, d row(d1 bigint, d2 double))), " +
+                "z array(array(row(p bigint, e row(e1 bigint, e2 varchar)))))");
+
+        // functions that are not outputting all subfields
```

Comment:
can you also add tests where the lambda doesn't access any subfields.  
Also, is it possible for a lambda to access a different column, could it access subfields from a struct that isn't one of the query arguments, and would we handle that properly (at least still return correct results)

Reply:
>can you also add tests where the lambda doesn't access any subfields.

Added.

>Also, is it possible for a lambda to access a different column, could it access subfields from a struct that isn't one of the query arguments, and would we handle that properly (at least still return correct results)

Yes, we will handle it properly. That is lambda descriptor for. When we attribute each lambda subfield, we check its name and compare with the name of the lambda argument. Moreover, different lambda arguments can relate to a different structs like with `ZIP_WITH` function. We also handle this correctly. 
There is only one catch here. When we accessing different column in lambda, additional call to `BIND` function will be added, which we do not support yet. As a result, we will give up on trying to collect the lambda subfields in this case and therefore include all of them.